pageserver/src/repository/inmemory.rs

```diff
@@ -0,0 +1,299 @@
+        debug!("put_truncation: {:?} at {}", relsize, lsn);
+        self.get_relfile(rel).put_truncation(lsn, relsize)
+    }
+
+    fn put_page_image(&self, tag: BufferTag, lsn: Lsn, img: Bytes) {
+        debug!("put_page_image: {:?} at {}", tag, lsn);
+        self.get_relfile(tag.rel).put_page_image(tag.blknum, lsn, img)
+    }
+
+    fn create_database(&self, _lsn: Lsn, _db_id: Oid, _tablespace_id: Oid, _src_db_id: Oid, _src_tablespace_id: Oid) -> Result<()> {
+
+        // TODO: Make a copy of everything in the old DB under new DB id
+        bail!("CREATE DATABASE not implemented");
+    }
+
+    /// Remember that WAL has been received and added to the page cache up to the given LSN
+    fn advance_last_valid_lsn(&self, lsn: Lsn) {
+        let old = self.last_valid_lsn.advance(lsn);
+
+        // Can't move backwards.
+        if lsn < old {
+            warn!(
+                "attempted to move last valid LSN backwards (was {}, new {})",
+                old, lsn
+            );
+        }
+    }
+
+    ///
+    /// Remember the (end of) last valid WAL record remembered in the page cache.
+    ///
+    /// NOTE: this updates last_valid_lsn as well.
+    ///
+    fn advance_last_record_lsn(&self, lsn: Lsn) {
+        // Can't move backwards.
+        let old = self.last_record_lsn.fetch_max(lsn);
+        assert!(old <= lsn);
+
+        // Also advance last_valid_lsn
+        let old = self.last_valid_lsn.advance(lsn);
+        // Can't move backwards.
+        if lsn < old {
+            warn!(
+                "attempted to move last record LSN backwards (was {}, new {})",
+                old, lsn
+            );
+        }
+    }
+
+    ///
+    /// Remember the beginning of valid WAL.
+    ///
+    /// TODO: This should be called by garbage collection, so that if an older
+    /// page is requested, we will return an error to the requestor.
+    /*
+    fn _advance_first_valid_lsn(&self, lsn: Lsn) {
+        // Can't overtake last_valid_lsn (except when we're
+        // initializing the system and last_valid_lsn hasn't been set yet.
+        let last_valid_lsn = self.last_valid_lsn.load();
+        assert!(last_valid_lsn == Lsn(0) || lsn < last_valid_lsn);
+
+        let old = self.first_valid_lsn.fetch_max(lsn);
+        // Can't move backwards.
+        assert!(lsn >= old);
+    }
+     */
+
+    fn init_valid_lsn(&self, lsn: Lsn) {
+        let old = self.last_valid_lsn.advance(lsn);
+        assert!(old == Lsn(0));
+        let old = self.last_record_lsn.fetch_max(lsn);
+        assert!(old == Lsn(0));
+        let old = self.first_valid_lsn.fetch_max(lsn);
+        assert!(old == Lsn(0));
+    }
+
+    fn get_last_valid_lsn(&self) -> Lsn {
+        self.last_valid_lsn.load()
+    }
+}
+
+impl InMemoryTimeline {
+
+    fn new(conf: &PageServerConf, timelineid: ZTimelineId) -> InMemoryTimeline {
+        InMemoryTimeline {
+            timelineid: timelineid,
```

Comment:
Don't need the right-hand side if the name is the same.

Run clippy to find these?